src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Snippets.java

```diff
@@ -0,0 +1,461 @@
+     *      ar[1] = elements[1];
+     *      ...
+     *      ar[pageSize-1] = elements[pageSize - 1];
+     *      methodNamePrefix1(ar);
+     *      return ar;
+     *   }
+     * and the last page will stop the chain and can be partial instead of full page size.
+     */
+    public static class ArraySnippetBuilder extends CollectionSnippetBuilder {
+        final MethodTypeDesc MTD_PageHelper;
+        final ClassDesc classDesc;
+        Snippet[] loadElementSnippets;
+
+        public ArraySnippetBuilder(ClassDesc elementType) {
+            super(elementType);
+            classDesc = elementType.arrayType();
+            MTD_PageHelper = MethodTypeDesc.of(classDesc, classDesc);
+        }
+
+        protected void fill(CodeBuilder cob, int fromIndex, int toIndex) {
+            for (var index = fromIndex; index < toIndex; index++) {
+                cob.dup()    // arrayref
+                   .loadConstant(index);
+                loadElementSnippets[index].emit(cob);  // value
+                cob.aastore();
+            }
+        }
+
+        private void invokePageHelper(CodeBuilder cob) {
+            // Invoke the first page, which will call next page until fulfilled
+            cob.loadConstant(loadElementSnippets.length)
+               .anewarray(elementType)
+               .invokestatic(ownerClassDesc, methodNamePrefix + "0", MTD_PageHelper);
+        }
+
+        private void newArray(CodeBuilder cob) {
+            cob.loadConstant(loadElementSnippets.length)
+               .anewarray(elementType);
+            fill(cob, 0, loadElementSnippets.length);
+        }
+
+        /**
+         * Generate helper methods to fill each page
+         */
+        private void setupHelpers() {
+            Objects.requireNonNull(clb);
+            Objects.requireNonNull(methodNamePrefix);
+            Objects.requireNonNull(ownerClassDesc);
+            var lastPageNo = new PagingContext(loadElementSnippets.length, pageSize).pageCount() - 1;
+            for (int pageNo = 0; pageNo <= lastPageNo; pageNo++) {
+                genFillPageHelper(pageNo, pageNo < lastPageNo);
+            }
+        }
+
+        // each helper function is T[] methodNamePrefix{pageNo}(T[])
+        // fill the page portion and chain calling to fill next page
+        private void genFillPageHelper(int pageNo, boolean hasNextPage) {
+            var fromIndex = pageSize * pageNo;
+            var toIndex = hasNextPage ? (fromIndex + pageSize) : loadElementSnippets.length;
+            clb.withMethodBody(methodNamePrefix + pageNo,
+                    MTD_PageHelper,
+                    ACC_STATIC,
+                    cob -> {
+                        cob.aload(0); // arrayref
+                        fill(cob, fromIndex, toIndex);
+                        if (hasNextPage) {
+                            cob.invokestatic(
+                                    ownerClassDesc,
+                                    methodNamePrefix + (pageNo + 1),
```

Comment:
```suggestion
                                    methodNamePrefix + "_" + (pageNo + 1),
```

Suggest to add a character e.g. `_` or `$` to separate the prefix and counter since the prefix may end with a number which is hard to see which part is prefix and which is page number.